text/0000-variadic-tuples.md

```diff
@@ -0,0 +1,1006 @@
+    ($a:ident,) => { $a };
+    ($a:ident, $($rest_a:ident,)+) => { last_type!($($rest_a,)+) };
+}
+```
+
+## Variadic tuple type
+
+### Declaration
+
+Variadic tuple types are always declared in a generic parameter group.
+
+There are two different syntaxes:
+
+1. `(..T)`: declare a single variadic tuple type identified by `T`
+2. `(..(T1, T2, ..., Tn))`: declare n variadic tuple types identified by `T1`, `T2`, ..., `Tn`, all these variadic tuple types have the same arity.
+
+Declaration examples:
+
+- `struct VariadicStruct<(..T1)>` : declares a struct with a variadic tuple type identified by `T1` in its generic parameters
+- `fn my_func<(..T1), (..T2)>()`: a function can have variadic tuple type parameters
+- `impl<(..Head)>`: is an implementation block that uses a variadic tuple type identified by `Head`
+- `impl<A, B, C, (.._Tail)>`:  same as above, but with other generic parameters
+- `impl<A, B, (..C), (..D)>`: there can be several variadic tuple types declared in a generic parameter group
+- `impl<A, B, (..(C, D)), (..E)>`: and we can mix both syntaxes
+
+Usage examples:
+
+```rust
+struct VariadicStruct<(..T)>
+VariadicStruct<(usize,)>                // => (..T) matches (usize,)
+VariadicStruct<(usize, bool)>   // => (..T) matches (usize, bool)
+
+impl <(..(T1, T2))> VariadicStruct<(..(T1, T2))> { ... }
+VariadicStruct::<((usize, bool),)>
+// (..(T1, T2)) matches ((usize, bool),)
+// (..T1) is (usize,)
+// (..T2) is (bool,)
+VariadicStruct::<((usize, bool), (String, i8))> // (..(T1, T2)) matches ((usize, bool), (String, i8))
+// (..T1) is (usize, String)
+// (..T2) is (bool, i8)
+```
+
+### Expansion
+
+The expansion syntax is: `..<expr(T1, T2, ..., Tn)>` where `<expr(T1, T2, ..., Tn)>` is an expression using the variadic tuple type identifiers `T1`, `T2`, ..., `Tn`.
+
+Note: The expression in an expansion form can be enclosed by parenthesis for clarity. Ex: `..(T: Clone)`.
+
+The expansion form is allowed in all places where a type is allowed and in `where` bounds.
+
+Examples:
+
+```rust
+type TuplesOfRef<'a, (..T)> = (..&'a T);
+TuplesOfRef<'b, (usize, bool)>; // = (&'b usize, &'b bool)
+
+struct MegaMap<(..(K, V))> {
+  maps: (..HashMap<K, V>),
+}
+// 
+// struct MegaMap<((usize, bool), (String, i8))> {
+//   maps: (HashMap<usize, bool>, HashMap<String, i8>),
+// }
+
+fn append<(..L), (..R)>(l: (..L), r: (..R)) -> (..L, ..R)
+where 
+    ..(L: 'static + Clone), 
+    ..(R: 'static + Clone), {
+    ...
+}
+
+//
+// fn append<(usize, Vec<bool>), (&'static str, u8, i16)>(
+//       l: (usize, Vec<bool>), 
+//       r: (&'static str, u8, i16)
+//   ) -> (usize, Vec<bool>, &'static str, u8, i16) {
+//   ...
+// }
+```
+
+Note: If an expansion syntax does not contains any variadic tuple type identifier, it resolves to the unit type `( )`.
+
+Note2: If an expansion syntax contains multiple variadic tuple type identifiers, they must all have been declared together with the syntax `( ..(T1, T2, ..., Tn))` to ensure they have the same arity.
```

Comment:
Yeah, these caveats should stay in the reference-level explanation. Use the examples to show this.

Also, I'm not sure why you're focusing on the case where the items in the variadic tuple are themselves variadic tuples. Seems like an obscure case to me.